src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -99,98 +114,138 @@ public NodeStateTable<T> AsCached()
                     compacted.Add(entry.AsCached());
                 }
             }
-            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true);
+            // When we're preparing a table for caching between runs, we drop the step information as we cannot guarantee the graph structure while also updating
```

Comment:
> The issue occurs in the following scenario: ...

It would be good to include this explanation in the code here as a comment.

Also, is there a way we can avoid this situation entirely? Naively, it feels like if the step info only included the filtered inputs (that is, the inputs from the previous node that were necessary), then the cached values should be correct even in this scenario. Sorry if I'm overlooking something fundamental.